core/trino-main/src/main/java/io/trino/connector/WorkerDynamicCatalogManager.java

```diff
@@ -205,9 +215,20 @@ private List<CatalogProperties> getMissingCatalogs(List<CatalogProperties> expec
     @Override
     public ConnectorServices getConnectorServices(CatalogHandle catalogHandle)
     {
-        CatalogConnector catalogConnector = catalogs.get(catalogHandle.getRootCatalogHandle());
+        Future<CatalogConnector> catalogConnector = catalogs.get(catalogHandle.getRootCatalogHandle());
         checkArgument(catalogConnector != null, "No catalog '%s'", catalogHandle.getCatalogName());
-        return catalogConnector.getMaterializedConnector(catalogHandle.getType());
+
+        try {
+            // This will synchronize while catalog is loading
+            return catalogConnector.get().getMaterializedConnector(catalogHandle.getType());
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        }
+        catch (ExecutionException e) {
+            throw new RuntimeException(e);
+        }
```

Comment:
I'd move the bulk of this to a new method `getCatalogConnector` and then this method would only be 
```java
return getCatalogConnector().getMaterializedConnector(catalogHandle.getType());